### PR TITLE
feat: [IEG-2717] CGN engagement banner

### PR DESCRIPTION
**Pull request by LeleDallas**
> [!NOTE]
> Depends on https://github.com/pagopa/io-services-metadata/pull/1170 

> [!TIP]
> To test checkout this [branch](https://github.com/pagopa/io-dev-api-server/pull/585) from dev server

## Short description
This pull request introduces the new CGN Discovery Banner feature to the services home screen, allowing eligible users to see a promotional banner for the Carta Giovani Nazionale (CGN). 

## List of changes proposed in this pull request
- Added the `CgnDiscoveryBanner` component, which displays a banner to eligible users who are not yet enrolled in CGN.
- Refactored `walletUpdate` action and saga to accept an `isRefresh` parameter, ensuring CGN details are only fetched when necessary (on refresh or first load), optimizing performance and reducing API calls.
- Added `useCgnEligibility` hook and improved `canAccessCgn` logic to robustly determine user eligibility based on date of birth or tax code, fixing edge cases for missing data
- Enhanced CGN details reducer to track whether details have already been fetched, preventing redundant network requests and enabling smarter UI decisions

## How to test
Verify that when a user does not have CGN, a banner is displayed in the Service tab to drive engagement.
Ensure that there are no duplicated api called to CGN status endpoint

## Preview 

https://github.com/user-attachments/assets/18c38819-c434-4791-ac88-54cfc3eacd4f

> Banner copy is outdated since the content is full remote

### CODE REVIEW

File: ts/features/bonus/cgn/utils/dates.ts [hunk 1]
```diff
@@ -7,18 +7,18 @@ type CgnUserAgeRange = "18-25" | "26-30" | "31-35" | "unrecognized";
...
-      return "26-30";
-    } else if (age > 17) {
-      return "18-25";
-    }
+  if (!profileBDay) {
+    return "unrecognized";
+  }
+  const date = new Date();
+  const birthDate = new Date(profileBDay);
+  const age = date.getFullYear() - birthDate.getFullYear();
```
Comment (emilio-dimari) [inline ts/features/bonus/cgn/utils/dates.ts]: doubt: ho notato un possibile edge case nel calcolo dell’età
(non introdotto da questa modifica).

Usando solo la differenza tra gli anni (`getFullYear()`), l’età potrebbe
risultare anticipata di 1 se il compleanno non è ancora passato, con
possibile impatto sui range.

**Review by emilio-dimari — APPROVED** [review]

File: ts/features/bonus/cgn/utils/dates.ts (hunk 1 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/utils/dates.ts]: As you can see from the git diff, I refactored the function to introduce an early return, avoiding unnecessary computations when the conditions are not met.

BTW, if we want to be more precise, we can use `ageFromDate` method listed below😄

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  const eligibleForCgn = useCgnEligibility();
+  const isCgnEnrolled = useIOSelector(isCgnEnrolledSelector);
+  const isBannerClosed = useIOSelector(isCgnDiscoveryBannerClosedSelector);
+
+  const cgnStatus = useIOSelector(cgnDetailSelector);
+  const cgnFetched = useIOSelector(isCgnAlreadyFetchedSelector);
+
+  useOnFirstRender(
+    () => {
+      if (isRemoteBannerEnabled && !cgnFetched && pot.isNone(cgnStatus)) {
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: What about grouping in a single variable with a proper meaningful name, all those conditions? 

`isRemoteBannerEnabled && !cgnFetched && pot.isNone(cgnStatus)`

In this way we improve code readability, what do you think?

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        dispatch(cgnDetails.request());
+      }
+    },
+    () =>
+      isRemoteBannerEnabled &&
+      !cgnFetched &&
+      pot.isNone(cgnStatus) &&
+      !isBannerClosed
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: As above

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  );
+
+  if (
+    !eligibleForCgn ||
+    !isRemoteBannerEnabled ||
+    !engagementBannerContent ||
+    isCgnEnrolled ||
+    isBannerClosed
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: as above

File: ts/features/bonus/cgn/store/actions/banners.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { ActionType, createStandardAction } from "typesafe-actions";
+
+export const setCgnDiscoveryBannerClosed = createStandardAction(
+  "CGN_DISCOVERY_BANNER_CLOSED"
+)<boolean>();
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/store/actions/banners.ts]: Do we really need a boolean payload for this action? The name already implies that the banner is being closed.
If this action is dispatched, shouldn’t we always set the Redux state to indicate the banner is closed (true)?
Is there any scenario where this action would be dispatched with a different meaning (e.g. false)?

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    >
+      <VSpacer size={16} />
+      <Banner
+        testID="itwDiscoveryBannerTestID"
+        title={engagementBannerContent.title?.[localeFallback]}
+        content={engagementBannerContent.description[localeFallback]}
+        action={
+          engagementBannerContent.action?.label[localeFallback] ??
+          "Attiva Carta Giovani"
+        }
```
[the remark this answers is not in the document]
Comment (LeleDallas) [reply ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx:89]: Solved in 3443c579214280b25fdba8c62429937cdd14855d

File: ts/features/bonus/cgn/store/reducers/details.ts [hunk 7]
```diff
@@ -57,6 +63,9 @@ export default reducer;
 export const cgnDetailSelector = (state: GlobalState) =>
   state.bonus.cgn.detail.information;
 
+export const isCgnAlreadyFetchedSelector = (state: GlobalState) =>
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/store/reducers/details.ts]:
```suggestion
export const isCgnDetailsAlreadyFetchedSelector = (state: GlobalState) =>
```

File: ts/features/bonus/cgn/utils/dates.ts [hunk 8]
```diff
@@ -45,3 +45,48 @@ export const getAccessibleExpirationDate = (
...
+const ageFromDate = (birthDate: Date, today: Date = new Date()): number => {
+  const age = today.getFullYear() - birthDate.getFullYear();
+  const hasBirthdayPassed =
+    today.getMonth() > birthDate.getMonth() ||
+    (today.getMonth() === birthDate.getMonth() &&
+      today.getDate() >= birthDate.getDate());
+  return hasBirthdayPassed ? age : age - 1;
+};
+
+const birthDateFromTaxCode = (cf: string): Date => {
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/utils/dates.ts:73]: What about adding tests for those two new functions? What do you think?

File: ts/features/bonus/cgn/utils/dates.ts [hunk 9]
```diff
@@ -45,3 +45,48 @@ export const getAccessibleExpirationDate = (
...
+  const month = MONTH_CODES[cf[8]];
+  const rawDay = Number.parseInt(cf.slice(9, 11), 10);
+  // Women have 40 added to their day of birth in the tax code
+  const day = rawDay > 40 ? rawDay - 40 : rawDay;
+  const currentYear = new Date().getFullYear() % 100;
+  const fullYear = (year <= currentYear ? 2000 : 1900) + year;
+  return new Date(fullYear, month, day);
+};
+
+export const canAccessCgn = (profile?: InitializedProfile) => {
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/utils/dates.ts:84]: As above

File: ts/features/bonus/cgn/utils/dates.ts (hunk 1 above)
Comment (emilio-dimari) [reply ts/features/bonus/cgn/utils/dates.ts]: The early return refactor is clear. My comment was only about the accuracy of the age calculation: using just the year difference can result in anticipating the age by 1 before the birthday.

A simple example: I was born in ’96 and my birthday is in June, so today 2026 - 1996 = 30, but in reality I’m still 29 and will turn 30 in a few months. 

From this perspective, something more precise like `ageFromDate` could be a good option 😄

Comment (emilio-dimari) [reply ts/features/bonus/cgn/utils/dates.ts]: Anyway, I know this is out of scope for this PR — just wanted to flag it in case we want to address it at some point.

File: ts/features/services/home/screens/ServicesHomeScreen.tsx
Comment (shadowsheep1) [inline ts/features/services/home/screens/ServicesHomeScreen.tsx:1]: Is this only an import "reordering" diff? I suggest  not to commit the file in those cases, to avoid to increase the review's cognitive complexity. What do you think? BTW, we can leave it here this time.

File: ts/features/bonus/cgn/utils/dates.ts (hunk 1 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/utils/dates.ts]: Makes sense to me! 
Addressed here 6763d83e2c9b9e97cafb36ef1fc1210ee13b00eb

File: ts/features/wallet/store/actions/index.ts [hunk 10]
```diff
@@ -5,8 +5,11 @@ import { WalletPreferencesActions } from "./preferences";
 
 /**
  * Action to trigger the update of the wallet screen content
+ * @param isRefresh - true when triggered by pull-to-refresh
  */
-export const walletUpdate = createStandardAction("WALLET_UPDATE")();
+export const walletUpdate = createStandardAction("WALLET_UPDATE")<{
+  isRefresh: boolean;
+}>();
```
Comment (shadowsheep1) [inline ts/features/wallet/store/actions/index.ts:12]: I see that we have changed this action, but we don't have any changes in any test files... maybe we are missing tests for this action? Can we get the chance to add some tests now?

File: ts/store/reducers/backendStatus/remoteConfig.ts [hunk 11]
```diff
@@ -143,6 +143,31 @@ export const isCGNEnabledAfterLoadSelector = createSelector(
   (isLoaded, isEnabled) => (isLoaded ? isEnabled : true)
 );
 
+export const isCGNDiscoveryBannerEnabledSelector = (state: GlobalState) =>
+  pipe(
+    state,
+    remoteConfigSelector,
+    O.map(config =>
+      isVersionSupported(
```
Comment (shadowsheep1) [inline ts/store/reducers/backendStatus/remoteConfig.ts]: I see you're not using `isMinAppVersionSupported`, but instead `isVersionSupported`, which could allow the flag to be enabled even for a `0.0.0.0` version and doesn’t provide the same input sanitization.

Is there a reason you prefer not to use isMinAppVersionSupported?

File: ts/store/reducers/backendStatus/remoteConfig.ts [hunk 12]
```diff
@@ -143,6 +143,31 @@ export const isCGNEnabledAfterLoadSelector = createSelector(
   (isLoaded, isEnabled) => (isLoaded ? isEnabled : true)
 );
 
+export const isCGNDiscoveryBannerEnabledSelector = (state: GlobalState) =>
```
Comment (shadowsheep1) [inline ts/store/reducers/backendStatus/remoteConfig.ts:146]: What about adding tests for this selector too?

File: ts/store/reducers/backendStatus/remoteConfig.ts [hunk 13]
```diff
@@ -143,6 +143,31 @@ export const isCGNEnabledAfterLoadSelector = createSelector(
...
+        Platform.OS === "ios"
+          ? config.cgn.show_cgn_engagement_banner?.min_app_version?.ios
+          : config.cgn.show_cgn_engagement_banner?.min_app_version?.android,
+        getAppVersion()
+      )
+    ),
+    O.getOrElse(() => false)
+  );
+
+export const engagementCGNDiscoveryBannerSelector = createSelector(
```
Comment (shadowsheep1) [inline ts/store/reducers/backendStatus/remoteConfig.ts:162]: As above

File: ts/features/bonus/cgn/store/reducers/banners.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+      return state;
+  }
+};
+
+const persistConfig: PersistConfig = {
+  key: "cgnBanners",
+  storage: AsyncStorage
+};
+
+export const cgnBannersPersistent = persistReducer(
```
Comment (shadowsheep1) [inline ts/features/bonus/cgn/store/reducers/banners.ts:39]: Since `logout` clears persisted data (see `createRootReducer`), a dismissed banner will reappear by default.

Is that intentional, or should this state survive logout?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: ts/features/bonus/cgn/utils/dates.ts (hunk 9 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/utils/dates.ts:84]: Addressed here be1eb314528fd75e8eaeb9b80bdd35695d22c8c7

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx (hunk 2 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: Makes sense!
Addressed here 53af34bddd2c9e7bf7408ab53828a904f99978e2

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx (hunk 3 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: Addressed here 53af34bddd2c9e7bf7408ab53828a904f99978e2

File: ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx (hunk 4 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/components/CgnDiscoveryBanner.tsx]: Addressed here 53af34bddd2c9e7bf7408ab53828a904f99978e2

File: ts/features/bonus/cgn/store/actions/banners.ts (hunk 5 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/store/actions/banners.ts]: In this case, there are no other scenarios where the action would be called with `true`.
If you agree, I’ll also update the action name in `closeCgnDiscoveryBanner` accordingly

File: ts/features/bonus/cgn/store/reducers/details.ts (hunk 7 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/store/reducers/details.ts]: Solved here d43ada01b33fa5a046e0756acffbf6b0e40d3a22

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 11 above)
Comment (LeleDallas) [reply ts/store/reducers/backendStatus/remoteConfig.ts]: I test it and you are right!
Addressed here 2f263aa0f7c74343143d801d6777ee78d1a02898

File: ts/features/services/home/screens/ServicesHomeScreen.tsx
Comment (LeleDallas) [reply ts/features/services/home/screens/ServicesHomeScreen.tsx:1]: The diff were created cause before I put the fetching logic inside this screen instead inside the banner component. 
If you want, I can revert these changes

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 12 above)
Comment (LeleDallas) [reply ts/store/reducers/backendStatus/remoteConfig.ts:146]: Added here 1c9f58a544bd8556426374f5697867552afd903c

File: ts/store/reducers/backendStatus/remoteConfig.ts (hunk 13 above)
Comment (LeleDallas) [reply ts/store/reducers/backendStatus/remoteConfig.ts:162]: Addressed here 35f9592307050881c217e8cd5b537b6361e2d5a2

File: ts/features/bonus/cgn/store/actions/banners.ts (hunk 5 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/store/actions/banners.ts]: Refactored here f1f828b658a68fe3c3a47dce1a72ac0b315e711e

File: ts/features/wallet/store/actions/index.ts (hunk 10 above)
Comment (LeleDallas) [reply ts/features/wallet/store/actions/index.ts:12]: Test suite added here be7a7e1ed552b844d81b2206d678acf5eb43aebb

File: ts/features/bonus/cgn/store/actions/banners.ts (hunk 5 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/store/actions/banners.ts]: Adjusted setter action here 595c8c1bf7d3051326ad8fa934f6a1d26b6d2c41

File: ts/features/bonus/cgn/utils/dates.ts (hunk 8 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/utils/dates.ts:73]: Looking at [codecov](https://app.codecov.io/gh/pagopa/io-app/pull/7977/blob/ts/features/bonus/cgn/utils/dates.ts), with latest test added, these functions seems to be covered

File: ts/features/bonus/cgn/store/reducers/banners.ts (hunk 14 above)
Comment (LeleDallas) [reply ts/features/bonus/cgn/store/reducers/banners.ts:39]: If I understand correctly, to fix this behavior I should store the CGN banner property object inside the `createRootReducer` function. Is that correct?

Comment (shadowsheep1) [reply ts/features/bonus/cgn/store/reducers/banners.ts:39]: Yep, and in the banner reducer you've to set its initial state. But, before doing all this stuff, let's consider if we want to address this scenario.

Comment (LeleDallas) [reply ts/features/bonus/cgn/store/reducers/banners.ts:39]: As we discuss in async, we can ignore this particular case

**Review by shadowsheep1 — APPROVED** [review]
LGTM

**Comment by emilio-dimari** [comment]
LGTM — tested locally ✅

**Comment by LeleDallas** [comment]
Closing this, as the banner will be relocated to a different screen as part of another development effort
